### ISSUE TITLE
cmake error when building shared lib without separated libs

### ISSUE THREAD

**Issue by vtorri** [lexbor/lexbor#61]
compiling on Windows only shared libn without separated : 

```
CMake Error at config.cmake:171 (set_property):
  set_property could not find TARGET lexbor-core.  Perhaps it has not yet
  been created.
Call Stack (most recent call first):
  CMakeLists.txt:178 (INCLUDE_MODULE_CONFIG)


-- Looking for ceil
-- Looking for ceil - found
-- Append module: core (1.3.1)
CMake Error at config.cmake:171 (set_property):
  set_property could not find TARGET lexbor-css.  Perhaps it has not yet been
  created.
Call Stack (most recent call first):
  CMakeLists.txt:178 (INCLUDE_MODULE_CONFIG)


-- Append module: css (0.1.0)
CMake Error at config.cmake:171 (set_property):
  set_property could not find TARGET lexbor-dom.  Perhaps it has not yet been
  created.
Call Stack (most recent call first):
  CMakeLists.txt:178 (INCLUDE_MODULE_CONFIG)


-- Append module: dom (1.2.1)
CMake Error at config.cmake:171 (set_property):
  set_property could not find TARGET lexbor-encoding.  Perhaps it has not yet
  been created.
Call Stack (most recent call first):
  CMakeLists.txt:178 (INCLUDE_MODULE_CONFIG)


-- Append module: encoding (2.0.1)
CMake Error at config.cmake:171 (set_property):
  set_property could not find TARGET lexbor-html.  Perhaps it has not yet
  been created.
Call Stack (most recent call first):
  CMakeLists.txt:178 (INCLUDE_MODULE_CONFIG)


-- Append module: html (1.4.0)
CMake Error at config.cmake:171 (set_property):
  set_property could not find TARGET lexbor-ns.  Perhaps it has not yet been
  created.
Call Stack (most recent call first):
  CMakeLists.txt:178 (INCLUDE_MODULE_CONFIG)


-- Append module: ns (1.2.0)
CMake Error at config.cmake:171 (set_property):
  set_property could not find TARGET lexbor-tag.  Perhaps it has not yet been
  created.
Call Stack (most recent call first):
  CMakeLists.txt:178 (INCLUDE_MODULE_CONFIG)


-- Append module: tag (1.2.0)
CMake Error at config.cmake:171 (set_property):
  set_property could not find TARGET lexbor-utils.  Perhaps it has not yet
  been created.
Call Stack (most recent call first):
  CMakeLists.txt:178 (INCLUDE_MODULE_CONFIG)


-- Append module: utils (0.2.1)
-- CFLAGS: -g3 -ggdb3 -Og -Wno-pedantic-ms-format -O2 -Wall -pedantic -pipe -std=c99
-- CXXFLAGS:  -O2
-- Feature ASAN: disable
-- Configuring incomplete, errors occurred!
See also "E:/Documents/programmes_x64/msys2/home/vtorri/gitroot_64/lexbor_vtorri/builddir/CMakeFiles/CMakeOutput.log".
```

it seems that config.cmake is included too early. But some macro are needed also (version for example) early.

so maybe separating config.cmake in 2 files, included at different locations could solve the probleM

**Comment by vtorri** [comment]
also, the diff of my modifications : 

```diff --git a/CMakeLists.txt b/CMakeLists.txt
index b75fce7..902d2b1 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -160,10 +160,12 @@ ENDFOREACH()
 ## First, need to add target for shared and static library
 IF(LEXBOR_BUILD_SHARED)
     add_library(${LEXBOR_LIB_NAME} SHARED ${LEXBOR_SOURCES})
+    set_property(TARGET ${LEXBOR_LIB_NAME} APPEND PROPERTY COMPILE_FLAGS "-DLEXBOR_BUILDING")
 ENDIF()

 IF(LEXBOR_BUILD_STATIC)
     add_library(${LEXBOR_LIB_NAME_STATIC} STATIC ${LEXBOR_SOURCES})
+    set_property(TARGET ${LEXBOR_LIB_NAME_STATIC} APPEND PROPERTY COMPILE_FLAGS "-DLEXBOR_STATIC")
 ENDIF()

 ## Load configurations from all modules. Print modules name and version.
diff --git a/config.cmake b/config.cmake
index bb6fbb4..e1e3029 100644
--- a/config.cmake
+++ b/config.cmake
@@ -166,8 +166,14 @@ MACRO(INCLUDE_MODULE_CONFIG pname module module_dir)
         set(DEPENDENCIES "")

         IF(EXISTS "${conf_path}")
-            set(CURRENT_LIB_NAME "${PROJECT_NAME}-${module}")
-            set(CURRENT_LIB_NAME_STATIC "${PROJECT_NAME}-${module}-static")
+            IF(LEXBOR_BUILD_SHARED)
+                set(CURRENT_LIB_NAME "${PROJECT_NAME}-${module}")
+                set_property(TARGET "${PROJECT_NAME}-${module}" APPEND PROPERTY COMPILE_FLAGS "-DLEXBOR_BUILDING")
+            ENDIF()
+
+            IF(LEXBOR_BUILD_STATIC)
+                set(CURRENT_LIB_NAME_STATIC "${PROJECT_NAME}-${module}-static")
+            ENDIF()

             include("${conf_path}")
         ENDIF()
@@ -310,7 +316,6 @@ MACRO(ADD_MODULE_LIBRARY type libname version_string major sources)
     ENDIF()

     set(postfix "")
-    set(dflags "-DLEXBOR_BUILDING")

     IF(NOT ${type} STREQUAL "")
         IF(${type} STREQUAL "STATIC")
diff --git a/examples/lexbor/css/CMakeLists.txt b/examples/lexbor/css/CMakeLists.txt
index 5e5c717..5dbd92a 100644
--- a/examples/lexbor/css/CMakeLists.txt
+++ b/examples/lexbor/css/CMakeLists.txt
@@ -13,8 +13,10 @@ set(syntax_tokenizer_parse_file_arg "${CMAKE_SOURCE_DIR}/test/files/lexbor/css/l
 ################
 ## Create executable
 #########################
-EXECUTABLE_LIST("" "${EXAMPLES_LEXBOR_CSS_SOURCES}" ${LEXBOR_LIB_NAME})
+IF(LEXBOR_BUILD_SHARED)
+    EXECUTABLE_LIST("" "${EXAMPLES_LEXBOR_CSS_SOURCES}" ${LEXBOR_LIB_NAME})

-IF(LEXBOR_BUILD_TESTS)
-    APPEND_TESTS("examples_lexbor_css_" "${EXAMPLES_LEXBOR_CSS_SOURCES}")
-ENDIF()
\ No newline at end of file
+    IF(LEXBOR_BUILD_TESTS)
+        APPEND_TESTS("examples_lexbor_css_" "${EXAMPLES_LEXBOR_CSS_SOURCES}")
+    ENDIF()
+ENDIF()
diff --git a/examples/lexbor/encoding/CMakeLists.txt b/examples/lexbor/encoding/CMakeLists.txt
index c28872e..c5b3e83 100644
--- a/examples/lexbor/encoding/CMakeLists.txt
+++ b/examples/lexbor/encoding/CMakeLists.txt
@@ -8,4 +8,6 @@ file(GLOB_RECURSE EXAMPLES_LEXBOR_ENCODING_SOURCES "*.c")
 ################
 ## Create executable
 #########################
-EXECUTABLE_LIST("" "${EXAMPLES_LEXBOR_ENCODING_SOURCES}" ${LEXBOR_LIB_NAME})
+IF(LEXBOR_BUILD_SHARED)
+    EXECUTABLE_LIST("" "${EXAMPLES_LEXBOR_ENCODING_SOURCES}" ${LEXBOR_LIB_NAME})
+ENDIF()
diff --git a/examples/lexbor/html/CMakeLists.txt b/examples/lexbor/html/CMakeLists.txt
index 94ec0e5..d396306 100644
--- a/examples/lexbor/html/CMakeLists.txt
+++ b/examples/lexbor/html/CMakeLists.txt
@@ -14,8 +14,10 @@ set(encoding_arg "${CMAKE_SOURCE_DIR}/test/files/lexbor/html/lexbor.html")
 ################
 ## Create executable
 #########################
-EXECUTABLE_LIST("" "${EXAMPLES_LEXBOR_HTML_SOURCES}" ${LEXBOR_LIB_NAME})
+IF(LEXBOR_BUILD_SHARED)
+    EXECUTABLE_LIST("" "${EXAMPLES_LEXBOR_HTML_SOURCES}" ${LEXBOR_LIB_NAME})

-IF(LEXBOR_BUILD_TESTS)
-    APPEND_TESTS("examples_lexbor_html_" "${EXAMPLES_LEXBOR_HTML_SOURCES}")
-ENDIF()
\ No newline at end of file
+    IF(LEXBOR_BUILD_TESTS)
+        APPEND_TESTS("examples_lexbor_html_" "${EXAMPLES_LEXBOR_HTML_SOURCES}")
+    ENDIF()
+ENDIF()
diff --git a/source/lexbor/core/def.h b/source/lexbor/core/def.h
index a8fb8ec..2d149ce 100755
--- a/source/lexbor/core/def.h
+++ b/source/lexbor/core/def.h
@@ -34,7 +34,11 @@

 #ifndef LEXBOR_STATIC
     #ifdef _WIN32
-        #define LXB_API __declspec(dllexport)
+        #ifdef LEXBOR_BUILDING
+            #define LXB_API __declspec(dllexport)
+        #else
+            #define LXB_API __declspec(dllimport)
+        #endif
     #elif (defined(__SUNPRO_C)  || defined(__SUNPRO_CC))
         #define LXB_API __global
     #else
@@ -49,13 +53,9 @@
 #endif

 #ifndef LEXBOR_STATIC
-    #ifdef _WIN32
-        #define LXB_EXTERN extern __declspec(dllimport)
-    #else
-        #define LXB_EXTERN extern
-    #endif
-#else
     #define LXB_EXTERN extern
+#else
+    #define LXB_EXTERN
 #endif

 #endif /* LEXBOR_DEF_H */
diff --git a/test/lexbor/core/CMakeLists.txt b/test/lexbor/core/CMakeLists.txt
index 60afc1a..77e49a3 100644
--- a/test/lexbor/core/CMakeLists.txt
+++ b/test/lexbor/core/CMakeLists.txt
@@ -13,5 +13,7 @@ file(GLOB_RECURSE TEST_LEXBOR_CORE_SOURCES "*.c")
 ################
 ## Create tests
 #########################
-EXECUTABLE_LIST("lexbor_core_" "${TEST_LEXBOR_CORE_SOURCES}" ${TEST_UNIT_LIB_NAME} ${LEXBOR_LIB_NAME})
-APPEND_TESTS("lexbor_core_" "${TEST_LEXBOR_CORE_SOURCES}")
\ No newline at end of file
+IF(LEXBOR_BUILD_SHARED)
+    EXECUTABLE_LIST("lexbor_core_" "${TEST_LEXBOR_CORE_SOURCES}" ${TEST_UNIT_LIB_NAME} ${LEXBOR_LIB_NAME})
+    APPEND_TESTS("lexbor_core_" "${TEST_LEXBOR_CORE_SOURCES}")
+ENDIF()
diff --git a/test/lexbor/css/CMakeLists.txt b/test/lexbor/css/CMakeLists.txt
index 443361a..f015a47 100644
--- a/test/lexbor/css/CMakeLists.txt
+++ b/test/lexbor/css/CMakeLists.txt
@@ -18,5 +18,7 @@ set(syntax_tokenizer_arg "${CMAKE_SOURCE_DIR}/test/files/lexbor/css/syntax/token
 ################
 ## Create tests
 #########################
-EXECUTABLE_LIST("lexbor_css_" "${TEST_LEXBOR_CSS_SOURCES}" ${TEST_UNIT_LIB_NAME} ${LEXBOR_LIB_NAME})
-APPEND_TESTS("lexbor_css_" "${TEST_LEXBOR_CSS_SOURCES}")
+IF(LEXBOR_BUILD_SHARED)
+    EXECUTABLE_LIST("lexbor_css_" "${TEST_LEXBOR_CSS_SOURCES}" ${TEST_UNIT_LIB_NAME} ${LEXBOR_LIB_NAME})
+    APPEND_TESTS("lexbor_css_" "${TEST_LEXBOR_CSS_SOURCES}")
+ENDIF()
diff --git a/test/lexbor/encoding/CMakeLists.txt b/test/lexbor/encoding/CMakeLists.txt
index a52361b..2e589cc 100644
--- a/test/lexbor/encoding/CMakeLists.txt
+++ b/test/lexbor/encoding/CMakeLists.txt
@@ -31,8 +31,10 @@ set(single_gb18030_arg "${CMAKE_SOURCE_DIR}/test/files/lexbor/encoding/gb18030_m
 ################
 ## Create tests
 #########################
-EXECUTABLE_LIST("lexbor_encoding_" "${TEST_LEXBOR_ENCODING_BUFFER_SOURCES}" ${TEST_UNIT_LIB_NAME} ${LEXBOR_LIB_NAME})
-APPEND_TESTS("lexbor_encoding_" "${TEST_LEXBOR_ENCODING_BUFFER_SOURCES}")
+IF(LEXBOR_BUILD_SHARED)
+    EXECUTABLE_LIST("lexbor_encoding_" "${TEST_LEXBOR_ENCODING_BUFFER_SOURCES}" ${TEST_UNIT_LIB_NAME} ${LEXBOR_LIB_NAME})
+    APPEND_TESTS("lexbor_encoding_" "${TEST_LEXBOR_ENCODING_BUFFER_SOURCES}")

-EXECUTABLE_LIST("lexbor_encoding_" "${TEST_LEXBOR_ENCODING_SINGLE_SOURCES}" ${TEST_UNIT_LIB_NAME} ${LEXBOR_LIB_NAME})
-APPEND_TESTS("lexbor_encoding_" "${TEST_LEXBOR_ENCODING_SINGLE_SOURCES}")
+    EXECUTABLE_LIST("lexbor_encoding_" "${TEST_LEXBOR_ENCODING_SINGLE_SOURCES}" ${TEST_UNIT_LIB_NAME} ${LEXBOR_LIB_NAME})
+    APPEND_TESTS("lexbor_encoding_" "${TEST_LEXBOR_ENCODING_SINGLE_SOURCES}")
+ENDIF()
diff --git a/test/lexbor/html/CMakeLists.txt b/test/lexbor/html/CMakeLists.txt
index 61b883b..2e00469 100644
--- a/test/lexbor/html/CMakeLists.txt
+++ b/test/lexbor/html/CMakeLists.txt
@@ -23,10 +23,12 @@ set(tree_builder_arg "${CMAKE_SOURCE_DIR}/test/files/lexbor/html/html5_test")
 ################
 ## Create tests
 #########################
-EXECUTABLE_LIST("lexbor_html_" "${TEST_LEXBOR_HTML_SOURCES}" ${TEST_UNIT_LIB_NAME} ${LEXBOR_LIB_NAME})
-APPEND_TESTS("lexbor_html_" "${TEST_LEXBOR_HTML_SOURCES}")
+IF(LEXBOR_BUILD_SHARED)
+    EXECUTABLE_LIST("lexbor_html_" "${TEST_LEXBOR_HTML_SOURCES}" ${TEST_UNIT_LIB_NAME} ${LEXBOR_LIB_NAME})
+    APPEND_TESTS("lexbor_html_" "${TEST_LEXBOR_HTML_SOURCES}")

-if(LEXBOR_BUILD_TESTS_CPP)
-    EXECUTABLE_LIST("lexbor_html_" "${TEST_LEXBOR_HTML_SOURCES_CPP}" ${TEST_UNIT_LIB_NAME} ${LEXBOR_LIB_NAME})
-    APPEND_TESTS("lexbor_html_" "${TEST_LEXBOR_HTML_SOURCES_CPP}")
-endif()
\ No newline at end of file
+    IF(LEXBOR_BUILD_TESTS_CPP)
+        EXECUTABLE_LIST("lexbor_html_" "${TEST_LEXBOR_HTML_SOURCES_CPP}" ${TEST_UNIT_LIB_NAME} ${LEXBOR_LIB_NAME})
+        APPEND_TESTS("lexbor_html_" "${TEST_LEXBOR_HTML_SOURCES_CPP}")
+    ENDIF()
+ENDIF()
diff --git a/test/lexbor/ns/CMakeLists.txt b/test/lexbor/ns/CMakeLists.txt
index 0bf7c8b..c81bfef 100644
--- a/test/lexbor/ns/CMakeLists.txt
+++ b/test/lexbor/ns/CMakeLists.txt
@@ -13,5 +13,7 @@ file(GLOB_RECURSE TEST_LEXBOR_NS_SOURCES "*.c")
 ################
 ## Create tests
 #########################
-EXECUTABLE_LIST("lexbor_ns_" "${TEST_LEXBOR_NS_SOURCES}" ${TEST_UNIT_LIB_NAME} ${LEXBOR_LIB_NAME})
-APPEND_TESTS("lexbor_ns_" "${TEST_LEXBOR_NS_SOURCES}")
+IF(LEXBOR_BUILD_SHARED)
+    EXECUTABLE_LIST("lexbor_ns_" "${TEST_LEXBOR_NS_SOURCES}" ${TEST_UNIT_LIB_NAME} ${LEXBOR_LIB_NAME})
+    APPEND_TESTS("lexbor_ns_" "${TEST_LEXBOR_NS_SOURCES}")
+ENDIF()
diff --git a/test/lexbor/tag/CMakeLists.txt b/test/lexbor/tag/CMakeLists.txt
index 6b1b529..6279b63 100644
--- a/test/lexbor/tag/CMakeLists.txt
+++ b/test/lexbor/tag/CMakeLists.txt
@@ -13,5 +13,7 @@ file(GLOB_RECURSE TEST_LEXBOR_TAG_SOURCES "*.c")
 ################
 ## Create tests
 #########################
-EXECUTABLE_LIST("lexbor_tag_" "${TEST_LEXBOR_TAG_SOURCES}" ${TEST_UNIT_LIB_NAME} ${LEXBOR_LIB_NAME})
-APPEND_TESTS("lexbor_tag_" "${TEST_LEXBOR_TAG_SOURCES}")
+IF(LEXBOR_BUILD_SHARED)
+    EXECUTABLE_LIST("lexbor_tag_" "${TEST_LEXBOR_TAG_SOURCES}" ${TEST_UNIT_LIB_NAME} ${LEXBOR_LIB_NAME})
+    APPEND_TESTS("lexbor_tag_" "${TEST_LEXBOR_TAG_SOURCES}")
+ENDIF()
diff --git a/test/unit/CMakeLists.txt b/test/unit/CMakeLists.txt
index 9ab9b8e..21dcdd4 100644
--- a/test/unit/CMakeLists.txt
+++ b/test/unit/CMakeLists.txt
@@ -23,18 +23,22 @@ file(GLOB_RECURSE TEST_UNIT_SOURCES "${TEST_UNIT_SOURCE_DIR}/*.c")
 ################
 ## Create Unit libraries
 #########################
-add_library(${TEST_UNIT_LIB_NAME_STATIC} STATIC ${TEST_UNIT_SOURCES})
-target_link_libraries(${TEST_UNIT_LIB_NAME_STATIC} ${LEXBOR_LIB_NAME})
-set_property(TARGET ${TEST_UNIT_LIB_NAME_STATIC} APPEND PROPERTY COMPILE_FLAGS "-DLEXBOR_STATIC")
+IF(LEXBOR_BUILD_STATIC)
+    add_library(${TEST_UNIT_LIB_NAME_STATIC} STATIC ${TEST_UNIT_SOURCES})
+    target_link_libraries(${TEST_UNIT_LIB_NAME_STATIC} ${LEXBOR_LIB_NAME})
+    set_property(TARGET ${TEST_UNIT_LIB_NAME_STATIC} APPEND PROPERTY COMPILE_FLAGS "-DLEXBOR_STATIC")
+ENDIF()

-add_library(${TEST_UNIT_LIB_NAME} SHARED ${TEST_UNIT_SOURCES})
-target_link_libraries(${TEST_UNIT_LIB_NAME} ${LEXBOR_LIB_NAME})
-set_property(TARGET ${TEST_UNIT_LIB_NAME} APPEND PROPERTY COMPILE_FLAGS "-DLEXBOR_BUILDING")
+IF(LEXBOR_BUILD_SHARED)
+    add_library(${TEST_UNIT_LIB_NAME} SHARED ${TEST_UNIT_SOURCES})
+    target_link_libraries(${TEST_UNIT_LIB_NAME} ${LEXBOR_LIB_NAME})
+    set_property(TARGET ${TEST_UNIT_LIB_NAME} APPEND PROPERTY COMPILE_FLAGS "-DLEXBOR_BUILDING")
+ENDIF()

-if (WIN32)
+IF(LEXBOR_BUILD_SHARED AND WIN32)
     add_custom_command(TARGET ${TEST_UNIT_LIB_NAME} POST_BUILD
         COMMAND ${CMAKE_COMMAND} -E copy_if_different "${CMAKE_CURRENT_BINARY_DIR}/lib${TEST_UNIT_LIB_NAME}.dll" $<TARGET_FILE_DIR:${LEXBOR_LIB_NAME}>/test/lexbor/core
         COMMAND ${CMAKE_COMMAND} -E copy_if_different "${CMAKE_CURRENT_BINARY_DIR}/lib${TEST_UNIT_LIB_NAME}.dll" $<TARGET_FILE_DIR:${LEXBOR_LIB_NAME}>/test/lexbor/html
         COMMAND ${CMAKE_COMMAND} -E copy_if_different "${CMAKE_CURRENT_BINARY_DIR}/lib${TEST_UNIT_LIB_NAME}.dll" $<TARGET_FILE_DIR:${LEXBOR_LIB_NAME}>/test/lexbor/ns
         COMMAND ${CMAKE_COMMAND} -E copy_if_different "${CMAKE_CURRENT_BINARY_DIR}/lib${TEST_UNIT_LIB_NAME}.dll" $<TARGET_FILE_DIR:${LEXBOR_LIB_NAME}>/test/lexbor/tag)
-endif (WIN32)
+ENDIF()
```

**Comment by lexborisov** [comment]
@vtorri 
I try to build project with your patch in MSVS `error LNK2001: unresolved external symbol lxb_encoding_res_map`
total: 217 errors

**Comment by vtorri** [comment]
i've not tried VS yet. I'll do once i get the issue fixed

**Comment by vtorri** [comment]
i've tried with VS 2019 : only 1 error that is related to syntax in a Visual Studio file, not 217 errors

**Comment by lexborisov** [comment]
@vtorri 

You can give more information how you run CMake and make in Windows? What parameters?

**Comment by vtorri** [comment]
I'm using MSYS2.

cross_toolchain.txt : 

```set(CMAKE_SYSTEM_NAME Windows)
set(CMAKE_SYSTEM_PROCESSOR AMD64)
set(CMAKE_C_COMPILER x86_64-w64-mingw32-gcc)
set(CMAKE_CXX_COMPILER x86_64-w64-mingw32-g++)
set(CMAKE_RC_COMPILER x86_64-w64-mingw32-windres)
```

cmake command : 

```#! /bin/bash

cd build_vs
cmake \
    -DCMAKE_TOOLCHAIN_FILE=cross_toolchain.txt \
    -DCMAKE_INSTALL_PREFIX=/opt/lexbor \
    -DLEXBOR_BUILD_STATIC=OFF \
    -DLEXBOR_BUILD_TESTS=ON \
    -DLEXBOR_BUILD_WITH_ASAN=ON \
    -DLEXBOR_C_FLAGS="-Wall -I../source -I../test -I../../test -I../../../test -I../../../../test -I../../../../../test -I../../source -I../../../source -I../../../../source" \
    -DLEXBOR_CXX_FLAGS="-Wall -I../source -I../test -I../../test -I../../../test -I../../../../test -I../../../../../test -I../../source -I../../../source -I../../../../source" \
    -G "Visual Studio 16" -A x64 \
    ..
```
then i double click on the VS solution file (lexbor.sln)
the Ctrl-Shift-b for the compilation

**Comment by lexborisov** [comment]
@vtorri 

Use `-DLEXBOR_BUILD_EXAMPLES=ON -DLEXBOR_BUILD_TESTS=ON` for generate VS Solution and try to build it.

**Comment by vtorri** [comment]
tests are already on

i'll try with examples later

**Comment by vtorri** [comment]
fixed inr PR #62

**Comment by vtorri** [comment]
fixed in PR #66 , #62 is closed
